### PR TITLE
Remove use of deprecated std::aligned_storage

### LINKED ISSUE #4806: `std::aligned_storage` is deprecated in C++23
I work on MSVC's C++ Standard Library, where we've recently implemented a C++23 paper that affects your project. (We regularly build many open-source projects to prevent compiler/library regressions, and this also allows us to notify projects of breaking changes that will affect them.)

The C++23 paper is https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p1413r3.pdf "Deprecate std::aligned_storage and std::aligned_union". We've merged a PR (https://github.com/microsoft/STL/pull/2583 ) that makes aligned_storage, aligned_storage_t, aligned_union, and aligned_union_t emit deprecation warnings when compiled in C++23 mode (/std:c++latest). This will be available in VS 2022 17.3 Preview 3. Although your project might not use /std:c++latest right now, we run such test coverage to identify issues that you might encounter in the future.

I searched and one line of affected code appears to be (there may be more):

https://github.com/KhronosGroup/SPIRV-Tools/blob/c26712784629f54c71b84b07bf161eabc0342ebe/source/util/small_vector.h#L470

(This emitted a warning when used by https://github.com/google/swiftshader , the project we built in our test harness.)

In general, the recommended replacement is to use the alignas keyword. We do provide an "escape hatch" macro to silence the deprecation warning, if necessary.

### PR DESCRIPTION
**Pull request by s-perron**
Fixes #4806.

### CODE REVIEW

File: source/util/small_vector.h [hunk 1]
```diff
@@ -461,15 +462,14 @@ class SmallVector {
   // The number of elements in |small_data_| that have been constructed.
   size_t size_;
 
+  // The actual data used to store the array elements.  It must never be used
+  // directly, but must only be accessed through |small_data_|.
+  alignas(T) std::array<std::byte, sizeof(T) * small_size> buffer;
```
Comment (Keenuts) [inline source/util/small_vector.h:467]: Should each element be aligned? Or is that a non-requirement?

```cpp
#include <array>
#include <cstddef>
#include <cstdio>
#include <stdint.h>

template <typename T>
void foo() {
  std::array<T, 2> array;
  void *ptr0 = (void*) &array[0];
  void *ptr1 = (void*) &array[1];
  printf("%p\n", ptr0);
  printf("%p\n", ptr1);
}

template <typename T>
void bar() {
  typename std::aligned_storage<sizeof(T), std::alignment_of<T>::value>::type array[2];
  void *ptr0 = (void*) &array[0];
  void *ptr1 = (void*) &array[1];
  printf("%p\n", ptr0);
  printf("%p\n", ptr1);
}

template <typename T>
void baz() {
  alignas(T) std::array<std::byte, sizeof(T) * 2> array;
  void *ptr0 = (void*) &array[0];
  void *ptr1 = (void*) &array[1];
  printf("%p\n", ptr0);
  printf("%p\n", ptr1);
}

int main() {
  foo<uint32_t>();
  bar<uint32_t>();
  baz<uint32_t>();
  return 0;
}
```

```
[nathan@Cassian ~]$ clang++ -std=c++17 main.cc -o main -O0 && ./main
0x7ffd54d56d00
0x7ffd54d56d04
0x7ffd54d56d00
0x7ffd54d56d04
0x7ffd54d56d00
0x7ffd54d56d01
```

Comment (s-perron) [reply source/util/small_vector.h:467]: You're right. I messed that up

**Comment by s-perron** [comment]
I should try to compile before I push a new commit.

File: source/util/small_vector.h [hunk 2]
```diff
@@ -461,14 +463,18 @@ class SmallVector {
...
+  // A type with the same alignment and size as T, but will is POD.
+  struct alignas(T) PodType {
+    std::array<std::byte, sizeof(T)> data;
+  };
 
   // The actual data used to store the array elements.  It must never be used
   // directly, but must only be accessed through |small_data_|.
-  typename std::aligned_storage<sizeof(T), std::alignment_of<T>::value>::type
-      buffer[small_size];
+  std::array<PodType, small_size> buffer;
```
Comment (Keenuts) [inline source/util/small_vector.h:473]: this fails to compile on gcc 12.2:
``small_vector.h: In instantiation of ‘spvtools::utils::SmallVector<T, small_size>::SmallVector() [with T = const spvtools::opt::analysis::Type*; long unsigned int small_size = 8]’:
small_vector.h:51:42: error: member ‘spvtools::utils::SmallVector<const spvtools::opt::analysis::Type*, 8>::buffer’ is used uninitialized [-Werror=uninitialized]
   51 |         small_data_(reinterpret_cast<T*>(buffer.data())),`

```

Fixed by adding `= { }` at the end.

Comment (s-perron) [reply source/util/small_vector.h]: I think this is fixed. I change the type of buffer to just a regular array instead of using std::array.

**Review by Keenuts — APPROVED** [review]

**Pull request merged**